Mpf get str fixups

bryant opened the pull request:
bug was introduced in 877104f58239fd19a4f3f5aed13bdd677c65f28b.

the intent here is to allocate enough space for gmpf_get_str to plop a string. if all digits are requested (n_digits == 0), we defer to gmpf_get_str itself to figure out how much space is needed (because the limb nonsense is too tedious to deal with) by passing nullptr to the first argument, as documented in the manual:

> If str is NULL, the result string is allocated using the current allocation function (see Custom Allocation). The block will be strlen(str)+1 bytes, that being exactly enough for the string and null-terminator.

if the number of digits requested is known (non-zero n_digits), then simply allocate a correctly sized cstring (unlike 877104f58239fd19a4f3f5aed13bdd677c65f28b, which allocates a 0-byte string).

to avoid memory leaks from repeated get_str(n_digits=0) calls, a wrapper is used to de-allocate on drop the gmp-allocated string.

with this patch, #24 passes.

bryant committed:
__gmpf_get_str is const on last arg.

from https://gmplib.org/manual/Converting-Floats.html : `Function: char
* mpf_get_str (char *str, mp_exp_t *expptr, int base, size_t n_digits,
const mpf_t op)`

bryant committed:
add wrapper for gmp-allocated strings

bryant committed:
properly call mpf_get_str

andersk commented on src/mpf.rs:
```diff
@@ -119,13 +120,24 @@ impl Mpf {
+            }
+        } else {
+            // From mpf/get_str.c.
+            let c_str = CString::new(vec![37; n_digits as usize + 2]).unwrap();
```
This intermediate `CString` is just overhead—why not use `Vec<c_char>::as_ptr` directly?

andersk commented on src/mpf.rs:
```diff
@@ -119,13 +120,24 @@ impl Mpf {
+            let c_str = CString::new(vec![37; n_digits as usize + 2]).unwrap();
+            unsafe {
+                let p = __gmpf_get_str(c_str.as_ptr(), exp, base, n_digits, &self.mpf);
+                CString::from_raw(p).to_str().unwrap().to_string()
```
Doesn’t this double-free the string, once via `c_str` and once via the `CString`? I think you want `CStr::from_ptr`.

andersk commented on src/ffi.rs:
```diff
@@ -1,7 +1,38 @@
+
+    pub fn to_str(&self) -> Result<&str, str::Utf8Error> {
+        let bytes: &[u8] = unsafe { slice::from_raw_parts(self.0, self.1) };
+        str::from_utf8(&bytes[..bytes.len() - 1])
```
`CStr::from_ptr` is more direct and avoids the separate `strlen`.

bryant replied:
no matter what, the length needs to be known in order to call the free func, so it's either strlen or CStr::to_bytes().len(). i think strlen communicates more clearly how the length param to free_func is computed, whereas the alternate method relies on knowing that CStr implicitly calls strlen behind the scenes.

bryant committed:
mpf_get_str's first arg is mut.

bryant committed:
rework the known n_digits case.

bryant committed:
pass null_mut

andersk commented on src/mpf.rs:
```diff
@@ -119,13 +120,26 @@ impl Mpf {
+        } else {
+            // From mpf/get_str.c.
+            let bytes = unsafe {
+                let mut bytes: Vec<u8> = uninitialized();
```
You need a `Vec` with uninitialized data, not an uninitialized `Vec`! I think `Vec::with_capacity` is the right thing.

bryant replied:
ugh, thanks for catching this.

andersk commented on src/mpf.rs:
```diff
@@ -119,13 +120,26 @@ impl Mpf {
+            // From mpf/get_str.c.
+            let bytes = unsafe {
+                let mut bytes: Vec<u8> = uninitialized();
+                let c_str: *mut c_char = transmute(bytes.as_mut_ptr());
```
No need for a worrying `transmute`—you can just `bytes.as_mut_ptr() as *mut c_char`.

bryant replied:
pretty sure that as_mut_ptr will return a * mut u8.

andersk commented on src/mpf.rs:
```diff
@@ -119,13 +120,26 @@ impl Mpf {
+                __gmpf_get_str(c_str, exp, base, n_digits, &self.mpf);
+                bytes
+            };
+            String::from_utf8(bytes).unwrap()
```
That includes the trailing NUL, right?

bryant replied:
yep. from mpf/get_str.c:

```c
  if (dbuf == 0)
    {
      /* We didn't get a string from the user.  Allocate one (and return
	 a pointer to it) with space for `-' and terminating null.  */
      alloc_size = n_digits + 2;
      dbuf = (char *) (*__gmp_allocate_func) (n_digits + 2);
```

bryant committed:
initialize the Vec with zeros.

this incorrectly created an uninitialized Vec, and using uninitialized
isn't worth it anyway.

bryant committed:
replace use of transmute with safe cast.

andersk commented on src/mpf.rs:
```diff
@@ -119,13 +120,26 @@ impl Mpf {
+        } else {
+            let bytes = unsafe {
+                // n_digits + 2 from mpf/get_str.c.
+                let mut bytes: Vec<u8> = vec![uninitialized(); n_digits as usize + 2];
```
This create a single uninitialized byte, then copies it `n_digits + 2` times, so you haven’t really saved any work.

I think what you want is
```rust
let mut buf = Vec::with_capacity(n_digits as usize + 2);
let ptr = __gmpf_get_str(buf.as_mut_ptr(), exp, base, n_digits, &self.mpf);
CStr::from_ptr(ptr).to_str().unwrap().to_string()
```

([prior art](https://github.com/rust-lang/rust/blob/1.18.0/src/libstd/sys/unix/os.rs#L490-L501))

bryant replied:
i think i'll revert to vec![0; n_digits + 2], which will call rust_allocate_zeroed which i'm guessing is more or less calloc. the with_capacity trick is neat but seems a little unsafe, and to be honest, i'm not sure that enough digits are requested for an appreciable amount of work to be saved.

andersk commented on src/mpf.rs:
```diff
@@ -119,13 +120,26 @@ impl Mpf {
+                __gmpf_get_str(c_str, exp, base, n_digits, &self.mpf);
+                bytes
+            };
+            String::from_utf8(bytes).unwrap()
```
No no, my point here was that `String::from_utf8` includes the trailing NUL in its _return value_.

bryant replied:
yeah, i realized what you meant a few minutes after hitting send. this branch should be tested anyway, and it shows up in the test.

bryant committed:
don't include null bytes from mpf_get_str.

bryant committed:
test get_str.

andersk commented on src/ffi.rs:
```diff
@@ -1,7 +1,38 @@
+        use std::ptr::null_mut;
+        unsafe {
+            let mut free_func: FreeFunc = mem::uninitialized();
+            __gmp_get_memory_functions(null_mut(), null_mut(), mem::transmute(&mut free_func));
```
This `transmute` is unnecessary because `&mut FreeFunc` converts implicitly to `*mut FreeFunc`.

fizyk20 commented:
Looks like this and #27 attempt to fix the same problem. I merged #27 before I had a look at your PR and now I think I might have done this too hastily. Your solution looks quite a bit more complex - do you see any obvious problems with how #27 attempts to do this?
